### ISSUE TITLE
Cannot get Certificate through istio

### ISSUE THREAD

**Issue by osaffer** [istio/istio#27244]
Hi,

I'm trying to access harbor through HTTPS protocol. My first test was about HTTP, and I did not have any problem.

kubectl version --short
Client Version: v1.19.0
Server Version: v1.19.0

istioctl version
client version: 1.7.0
control plane version: 1.7.0
data plane version: 1.7.0 (5 proxies)

Architecture, I have a Haproxy configured in PassThrough, already used with a nginx ingress controller, and here, no problem with HTTPS.

If I expose the Service as NodePort, 
openssl s_client workernode:32001 will show me the certificate.

When I do an openssl s_client VIP:443 the certificate is not showed.

HAProxy redirect VIP to my workers nodes by using the NodePort.

This is my GW section:

apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: harbor-gateway
  namespace: harbor
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "*"
    tls:
      httpsRedirect: true
  - port:
      number: 443
      name: https
      protocol: HTTPS
    hosts:
    - "*"
    tls:
      mode: SIMPLE
      credentialName: server-certs
      #serverCertificate: /opt/Harbor/server.crt
      #privateKey: /opt/Harbor/serverk.key

I even tried to use mode: PASSTHROUGH

But the certificate is never taken. 

Thanks for your help.

**Comment by howardjohn** [comment]
Ca nyou try running:

```
istioctl pc listener INGERSS_POD
istioctl pc secret
istioctl pc route
```

These will give useful info. If not, get a full config dump and post here: https://github.com/istio/istio/wiki/Troubleshooting-Istio

**Comment by osaffer** [comment]
Hi,

I am sorry for the late reply, 

I think it is more a Haproxy configuration issue. 
HAproxy is configured as passthrough: 
frontend https_istio
   mode http
   option forwardfor
   bind VIP:443 transparent
   default_backend https_istio_back

backend https_istio_back
    mode http
    option log-health-checks
    balance roundrobin
    timeout connect 10s
    timeout server 1m
    server host1 IP1:30602 check fall 3 rise 2 
    server host2 IP2:30602 check fall 3 rise 2
    server host3 IP3:30602 check fall 3 rise 2


 istioctl pc listener istio-ingressgateway-6785d64647-dc55l -n istio-system
ADDRESS PORT  MATCH DESTINATION
0.0.0.0 5432  ALL   Cluster: outbound|5432||postgres-b2g-postgresql-ha-pgpool.b2g.svc.cluster.local
0.0.0.0 8080  ALL   Route: http.80
0.0.0.0 15021 ALL   Inline Route: /healthz/ready*
0.0.0.0 15090 ALL   Inline Route: /stats/prometheus*

istioctl pc secret istio-ingressgateway-6785d64647-dc55l -n istio-system
RESOURCE NAME     TYPE           STATUS     VALID CERT     SERIAL NUMBER                               NOT AFTER                NOT BEFORE
default           Cert Chain     ACTIVE     true           163095856524359579433041700859108685722     2020-09-16T00:55:27Z     2020-09-15T00:55:27Z
ROOTCA            CA             ACTIVE     true           21767561330579796279867322371727544403      2030-08-31T13:09:03Z     2020-09-02T13:09:03Z

istioctl pc route istio-ingressgateway-6785d64647-dc55l -n istio-system
NOTE: This output only contains routes loaded via RDS.
NAME        DOMAINS     MATCH                  VIRTUAL SERVICE
http.80     *           /kiali*                kiali-vs.istio-system
http.80     *           /apache/*              apache-vs.default
            *           /healthz/ready*
            *           /stats/prometheus*

Regards,

**Comment by howardjohn** [comment]
Looks like there are no listeners for 443 at all. Do you have any virtual services defined for the gateway?

If I just apply the GW you had in the original issue, I get the same result. Adding a virtual service adds the listener. See:

```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: harbor-gateway
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "*"
    tls:
      httpsRedirect: true
  - port:
      number: 443
      name: https
      protocol: HTTPS
    hosts:
    - "*"
    tls:
      mode: SIMPLE
      credentialName: server-certs
      #serverCertificate: /opt/Harbor/server.crt
      #privateKey: /opt/Harbor/serverk.key
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: bookinfo
spec:
  hosts:
  - "*"
  gateways:
  - harbor-gateway
  http:
  - match:
    - uri:
        exact: /productpage
    route:
    - destination:
        host: productpage
        port:
          number: 9080
```

for example

**Comment by osaffer** [comment]
Hi,

kubectl  get vs,gw,dr -n harbor
NAME                                           GATEWAYS             HOSTS   AGE
virtualservice.networking.istio.io/harbor-vs   ["harbor-gateway"]   ["*"]   6m2s
virtualservice.networking.istio.io/portal-vs   ["portal-gateway"]   ["*"]   6m2s

NAME                                         AGE
gateway.networking.istio.io/harbor-gateway   6m2s
gateway.networking.istio.io/portal-gateway   6m2s

NAME                                                   HOST                                     AGE
destinationrule.networking.istio.io/harbor-dr          harbor.harbor.svc.cluster.local          70s
destinationrule.networking.istio.io/harbor-portal-dr   harbor-portal.harbor.svc.cluster.local   70s

Indeed, no Listener.

apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: harbor-gateway
  namespace: harbor
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "*"
    tls:
      httpsRedirect: true
  - port:
      number: 443
      name: https
      protocol: HTTPS
    hosts:
    - "*"
    tls:
      mode: SIMPLE
      credentialName: server-certs
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: harbor-vs
  namespace: harbor
spec:
  hosts:
  - "*"
  gateways:
  - harbor-gateway
  http:
  - match:
    - uri:
        prefix: /api/
    rewrite:
      uri: /api/
    route:
    - destination:
        host: harbor.harbor.svc.cluster.local
        port:
          number: 443
  - match:
    - uri:
        prefix: /service/
    rewrite:
      uri: /service/
    route:
    - destination:
        host: harbor.harbor.svc.cluster.local
        port:
          number: 443
  - match:
    - uri:
        prefix: /v2/
    rewrite:
      uri: /v2/
    route:
    - destination:
        host: harbor.harbor.svc.cluster.local
        port:
          number: 443
  - match:
    - uri:
        prefix: /chartrepo/
    rewrite:
      uri: /chartrepo/
    route:
    - destination:
        host: harbor.harbor.svc.cluster.local
        port:
          number: 443
  - match:
    - uri:
        prefix: /c/
    rewrite:
      uri: /c/
    route:
    - destination:
        host: harbor.harbor.svc.cluster.local
        port:
          number: 443

---

apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: portal-gateway
  namespace: harbor
spec:
  selector:
istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "*"
    tls:
      httpsRedirect: true
  - port:
      number: 443
      name: https
      protocol: HTTPS
    hosts:
    - "*"
    tls:
      mode: SIMPLE
      credentialName: server-certs

---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: portal-vs
  namespace: harbor
spec:
  hosts:
  - "*"
  gateways:
  - portal-gateway
  http:
  - match:
    - uri:
        prefix: /
    rewrite:
      uri: /
    route:
    - destination:
        host: harbor-portal.harbor.svc.cluster.local
        port:
          number: 443
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: harbor-dr
  namespace: harbor
spec:
  host: harbor.harbor.svc.cluster.local
  trafficPolicy:
    tls:
      mode: SIMPLE
      credentialName: server-certs
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: harbor-portal-dr
  namespace: harbor
spec:
  host: harbor-portal.harbor.svc.cluster.local
  trafficPolicy:
    tls:
      mode: SIMPLE
      credentialName: server-certs

Logs in istio gateway pod:

020-09-17T15:19:29.150741Z	info	Envoy command: [-c etc/istio/proxy/envoy-rev0.json --restart-epoch 0 --drain-time-s 45 --parent-shutdown-time-s 60 --service-cluster istio-ingressgateway --service-node router~172.20.1.193~istio-ingressgateway-6785d64647-xj2zn.istio-system~istio-system.svc.cluster.local --local-address-ip-version v4 --log-format-prefix-with-location 0 --log-format %Y-%m-%dT%T.%fZ	%l	envoy %n	%v -l warning --component-log-level misc:error]
2020-09-17T15:19:29.191634Z	warning	envoy runtime	Unable to use runtime singleton for feature envoy.reloadable_features.activate_fds_next_event_loop
2020-09-17T15:19:29.249383Z	warning	envoy config	StreamAggregatedResources gRPC config stream closed: 14, no healthy upstream
2020-09-17T15:19:29.249614Z	warning	envoy config	Unable to establish new stream
2020-09-17T15:19:29.259907Z	info	sds	resource:default new connection
2020-09-17T15:19:29.260088Z	info	sds	Skipping waiting for gateway secret
2020-09-17T15:19:29.605758Z	info	cache	Root cert has changed, start rotating root cert for SDS clients
2020-09-17T15:19:29.605857Z	info	cache	GenerateSecret default
2020-09-17T15:19:29.606480Z	info	sds	resource:default pushed key/cert pair to proxy
2020-09-17T15:19:29.608813Z	warning	envoy main	there is no configured limit to the number of allowed active connections. Set a limit via the runtime key overload.global_downstream_max_connections
2020-09-17T15:19:29.800152Z	info	sds	resource:ROOTCA new connection
2020-09-17T15:19:29.800348Z	info	sds	Skipping waiting for gateway secret
2020-09-17T15:19:29.800410Z	info	cache	Loaded root cert from certificate ROOTCA
2020-09-17T15:19:29.800645Z	info	sds	resource:ROOTCA pushed root cert to proxy
2020-09-17T15:19:30.932018Z	info	Envoy proxy is ready
2020-09-17T15:29:54.931738Z	info	sds	resource:server-certs-cacert new connection
2020-09-17T15:29:54.934744Z	warn	secretfetcher	Cannot find secret server-certs-cacert, searching for fallback secret gateway-fallback
2020-09-17T15:29:54.934766Z	error	secretfetcher	cannot find secret server-certs-cacert and cannot find fallback secret gateway-fallback
2020-09-17T15:29:54.934772Z	warn	cache	resource:server-certs-cacert SecretFetcher cannot find secret server-certs-cacert from cache
2020-09-17T15:29:54.934789Z	warn	sds	resource:server-certs-cacert waiting for gateway secret for proxy "router~172.20.1.193~istio-ingressgateway-6785d64647-xj2zn.istio-system~istio-system.svc.cluster.local"


Thank you

**Comment by osaffer** [comment]
Hi, 

Sorry for late reply, I had time to work on this problem, and it is solved now ! 

I have just performed a simple test with bookinfo.

I have upgraded istio to 1.7.2

 **istioctl pc secret istio-ingressgateway-56b8d79bfc-drmcb -n istio-system**
RESOURCE NAME           TYPE           STATUS     VALID CERT     SERIAL NUMBER                                        NOT AFTER                NOT BEFORE
**bookinfo-credential**     CA             ACTIVE     true           533739060012167978178216818319304990897580334194     2021-09-24T14:16:46Z     2020-09-24T14:16:46Z
default                 Cert Chain     ACTIVE     true           136136904787137501211906138307650109552              2020-09-25T14:04:11Z     2020-09-24T14:04:11Z
ROOTCA                  CA             ACTIVE     true           21767561330579796279867322371727544403               2030-08-31T13:09:03Z     2020-09-02T13:09:03Z

 **istioctl pc route istio-ingressgateway-56b8d79bfc-drmcb -n istio-system**
NOTE: This output only contains routes loaded via RDS.
NAME                                        DOMAINS             MATCH                  VIRTUAL SERVICE
**https.443.https.bookinfogateway.default**     myhostname     /*                     bookinfo.default
                                            *                   /healthz/ready*        
                                            *                   /stats/prometheus*     

istioctl pc listener istio-ingressgateway-56b8d79bfc-drmcb  -n istio-system
ADDRESS PORT  MATCH                DESTINATION
0.0.0.0 5432  ALL                  Cluster: outbound|5432||postgres-b2g-postgresql-ha-pgpool.b2g.svc.cluster.local
**0.0.0.0 8443  SNI: myhostname Route: https.443.https.bookinfogateway.default**
0.0.0.0 15021 ALL                  Non-HTTP/Non-TCP
0.0.0.0 15090 ALL                  Non-HTTP/Non-TCP

Ok now I have listener, before there was no listener because I think I used "*" instead of a hostname 

Now the problem is that I cannot reach the certificate when I do an openssl s_client myhostname:443

For information In front end I have configured  a LoadBalancer haproxy in passthrough mode. 

This is my config in order to make it work ! 
In global section, add: 

ssl-default-bind-options no-sslv3
tune.ssl.default-dh-param 2048


frontend https_istio
   mode tcp
   bind VIP:443 transparent
   default_backend https_istio_back

backend https_istio_back
    mode tcp
    option log-health-checks
    balance roundrobin
    timeout connect 10s
    timeout server 1m
    server worker1 IP:30602 check fall 3 rise 2
    server worker2 IP:30602 check fall 3 rise 2
    server worker3 IP:30602 check fall 3 rise 2

Regards,
